Exclusively OpenXR for Android/PCVR

mikeskydev opened the pull request:
This PR implements the early stages of https://github.com/mikeskydev/unity-openxr-extensions, moving away from vendor specific XR SDKs towards using OpenXR for as many features as possible.

TODO:
- [x] Fix up AndroidManifest; some things are implemented in the package, others *should* be.
- [x] Test Pico a bit more thoroughly.
- [ ] Re-orient pico controllers

Out of scope:
- Switch MX Ink to OpenXR #942.

Test Matrix Nov '25:
- [x] Quest 3
- [ ] Quest 1
- [x] Pico 4/Ultra
- [ ] Windows

This removes dependency on the Oculus Core SDK and the Pico SDK, allowing us to use the same codepath. Annoyingly our package name is still different on Oculus, so we have to do a separate build. @mikeage could we consider using apktool to hack our own package as a post-build step, rather than building twice?

Oculus Passthrough is delivered via the new upm package, and as a plus we get passthrough on Pico and Quest PCVR!

Oculus Platform is required for age check (and some MR features that are currently commented out), but we check which platform we're running on before initializing, otherwise it can crash non-oculus android devices. I will reimplement the MR features in my OpenXR package at a later date.

Wave is removed (never actually used, and we can do it properly with OpenXR), as is the Pimax build, getting our current builds down to 6.

Closes #447

mikeskydev committed:
openxr 1.13, composition layers

mikeskydev committed:
mikeskydev openxr extensions

mikeskydev committed:
update meta platform to remove meta core deps

mikeskydev committed:
goodbye oculus

mikeskydev committed:
first pass remove oculus/pico defines

mikeskydev committed:
might as well ditch wave too

mikeskydev committed:
up to date sdks/platforms

mikeskydev committed:
wrap more oculus

mikeskydev committed:
remove pimax

mikeskydev committed:
remove oculus controls

mikeskydev committed:
remove redundant input manager

mikeskydev committed:
fix initialization, openxr stage,

mikeskydev committed:
oculus platform tidy up

mikeskydev committed:
prevent platform crash on non-quest

mikeskydev committed:
Redefine passthrough

mikeskydev committed:
Rework build

[CI BUILD]

mikeskydev committed:
junk

mikeskydev committed:
remove buildtime configuration of CN variant

[CI BUILD]

mikeskydev committed:
format AndroidUtils

[CI BUILD]

mikeskydev committed:
extensions version update

[CI BUILD]

mikeskydev committed:
no splash for now

mikeskydev committed:
Pico controller, CN support re added

[CI BUILD]

mikeskydev committed:
Merge branch 'main' into feature/pure-openxr

# Conflicts:
#	Assets/Oculus/OculusProjectConfig.asset
#	Assets/XR/Settings/Open XR Package Settings.asset

mikeage commented:
apktool may be a drop problematic; we have to make sure it's signed properly. But it's probably possible. Maybe a followup PR?

mikeage commented on .github/workflows/build.yml:
```diff
@@ -1222,9 +1082,8 @@ jobs:
-            # The order here matters, because the Chinese build has a slightly higher version code due to the suffix of 1 vs 0
             ./pico-cli upload-build --app-id $PICO_APP_ID --app-secret $PICO_APP_SECRET --region noncn --apk OpenBrush_Pico_$VERSION.apk --channel 2 --notes-en "Version $VERSION" --device 'PICO Neo3,PICO Neo3 Pro,PICO Neo3 Eye,PICO 4'
-            ./pico-cli upload-build --app-id $PICO_APP_ID --app-secret $PICO_APP_SECRET --region noncn --apk OpenBrush_Pico_CN_$VERSION.apk --channel 1 --notes-en "Version $VERSION" --device 'PICO Neo3,PICO Neo3 Pro,PICO Neo3 Eye,PICO 4'
+            ./pico-cli upload-build --app-id $PICO_APP_ID --app-secret $PICO_APP_SECRET --region cn --apk OpenBrush_Pico_$VERSION.apk --channel 1 --notes-en "Version $VERSION" --device 'PICO Neo3,PICO Neo3 Pro,PICO Neo3 Eye,PICO 4'
```
I fear that this will fail; IIRC, we can't publish the same version code to both CN and non-CN.

mikeskydev replied:
Ah, that's incredibly dumb if so. Can we give it a try? The code I'm using to unify the builds is from Pico's official Unreal documentation, so it implies the same build should be supported?

mikeage replied:
Sure. Whenever we merge this, just be on call for followups with failed publishes, including possibly oculus (because of course they will), which is something we can't easily test

mikeskydev replied:
Yep, I will also try taking a build and manually uploading to a test app I have before merging

mikeage replied:
ok, just don't upload to our regular build. The way they do channels is a bit weird, but the upshot is that you can mess things up by pushing random android build IDs

mikeskydev replied:
![image](https://github.com/user-attachments/assets/ba9c9267-d9e5-4db2-a1a7-01999aabf062)

mikeage replied:
I hope that's accurate :-)

mikeage commented on Assets/Plugins/Android/AndroidManifest.xml:
```diff
@@ -16,10 +16,19 @@
+
+    <meta-data android:name="com.oculus.vr.focusaware" android:value="true"/>
+    <meta-data android:name="com.oculus.supportedDevices" android:value="quest|quest2" />
+    <meta-data android:name="pvr.app.type" android:value="vr" />
```
did you mean to move this outside of `<application>`?

mikeskydev replied:
I actually meant to remove it entirely, my plugin can handle it now! Keep this open as a reminder

mikeskydev commented:
Yeah it can be a followup, I'm hoping we can make the change then re-sign as we have our original keys

mikeskydev committed:
Merge tag '2.9.14' into feature/pure-openxr

mikeskydev committed:
control xr manifest bits via plugin

mikeskydev committed:
controller detection

mikeskydev committed:
why are feature settings so buggy

mikeskydev committed:
init android manager on start

github-bot committed:
Merge branch 'main' into feature/pure-openxr

# Conflicts:
#	Assets/Plugins/Android/AndroidManifest.xml
#	Assets/Scripts/GUI/LightingPopUpWindow.cs
#	Assets/Scripts/Multiplayer/MultiplayerManager.cs

mikeskydev committed:
Stable OpenXR settings? finally?

Had to delete my library...

mikeskydev committed:
Merge branch 'main' into feature/pure-openxr

mikeskydev committed:
pico offsets?

[CI BUILD]

mikeskydev committed:
Merge branch 'main' into feature/pure-openxr

andybak commented:
EDIT - looking at the todo above - it seems there's only one remaining item other than the MX Ink. However - I strongly suspect the missing todo item is "do a ton of testing and find lots of things we've overlooked"!

smitdylan2001 commented:
In general I couldn't find passthrough options in the version on the Pico
store (unless if that's recently added)

On Thu, Sep 4, 2025, 16:41 eeease ***@***.***> wrote:

> *eeease* left a comment (icosa-foundation/open-brush#798)
> <https://github.com/icosa-foundation/open-brush/pull/798#issuecomment-3254064332>
>
> Hi all! Amazing work on this, thanks for keeping it open for others to
> learn from and contribute to. I was coincidentally just working on making
> the move to pure OpenXR for Quest and PICO, so a lot of this confirms
> conclusions I came.
>
> At the moment I'm particularly interested in OpenXR Passthrough on PICO.
>
> I cloned the Open Brush repo, switched to this branch, and opened the
> project in Unity. Can you let me know how you're enabling passthrough on
> PICO? I searched through all of the menu options but didn't see a button
> in-app. I haven't *extensively* looked through all the code yet but I
> figured I'd ask to save some time in case I'm missing something.
>
> Note: I tried both in Editor and in a standalone Windows build. I'm using
> a PICO Business device so using PICO Business Connect; picostreaming is the
> preferred OpenXR runtime on my computer atm.
>
> Thanks!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/icosa-foundation/open-brush/pull/798#issuecomment-3254064332>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJ5EZXGILQEGFZXTH5JTJAL3RBFS3AVCNFSM6AAAAACE6BE67CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTENJUGA3DIMZTGI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

andybak committed:
dummy commit

andybak committed:
[CI BUILD]

mikeskydev committed:
Merge branch 'main' into feature/pure-openxr

# Conflicts:
#	.github/workflows/build.yml
#	Assets/Oculus/OculusProjectConfig.asset
#	Assets/Prefabs/NoPeekingCamera.prefab
#	Assets/Scripts/Multiplayer/MultiplayerManager.cs
#	Assets/Scripts/VrSdk.cs
#	Assets/XR/Settings/Oculus Settings.asset
#	Assets/XR/Settings/OpenXRPackageSettings.asset
#	Packages/manifest.json

mikeskydev committed:
redo 7defcfce4d977654345d0fe9d64257a50eb6dd08

mikeage commented on .github/workflows/build.yml:
```diff
@@ -1366,9 +1261,8 @@ jobs:
-            # The order here matters, because the Chinese build has a slightly higher version code due to the suffix of 1 vs 0
             ./pico-cli upload-build --app-id $PICO_APP_ID --app-secret $PICO_APP_SECRET --region noncn --apk OpenBrush_Pico_$VERSION.apk --channel 2 --notes-en "Version $VERSION" --device 'PICO Neo3,PICO Neo3 Pro,PICO Neo3 Eye,PICO 4,PICO 4 Ultra,PICO 4 Pro'
-            ./pico-cli upload-build --app-id $PICO_APP_ID --app-secret $PICO_APP_SECRET --region noncn --apk OpenBrush_Pico_CN_$VERSION.apk --channel 1 --notes-en "Version $VERSION" --device 'PICO Neo3,PICO Neo3 Pro,PICO Neo3 Eye,PICO 4,PICO 4 Ultra,PICO 4 Pro'
+            ./pico-cli upload-build --app-id $PICO_APP_ID --app-secret $PICO_APP_SECRET --region cn --apk OpenBrush_Pico_$VERSION.apk --channel 1 --notes-en "Version $VERSION" --device 'PICO Neo3,PICO Neo3 Pro,PICO Neo3 Eye,PICO 4,PICO 4 Ultra,PICO 4 Pro'
```
Is this correct?

mikeage replied:
we really need to make sure we understand how this stuff works. Channel 1 is the China Store, Channel 2 is non-China, but there's also region, and... I don't know

mikeage replied:
I'm not sure. Since we're outside of China, I think we need to use --region noncn, even when we upload to --channel 1 (and certainly for --channel 2)

mikeskydev replied:
Oh, I get you. It'll be doing something lime be changing the endpoint target for the upload, that wasn't clear.

mikeage reviewed (commented):
A couple of questions:

1. For Q1, we had to force a few extra things in the manifest. I'm not sure if we still need those since we're not longer pushing to the store
2. ~~We still need a solution for the different package names.~~ edit: I misread the build.yml
3. Does Pimax still work?
4. I think we should test manual uploads to a different release channel on the platforms that support it (i.e., Q2+ and Pico) just to make sure its happy. We need to make sure that this does not go to the beta or live channels, though, lest it cause problems with downgrades.

Also note that I'm going on vacation for a week or so starting on Thursday :-)

mikeage commented:
And it looks like the Github publish might also be broken

mikeskydev committed:
fix target downloads

mikeage commented:
A few comments:

what about the Rift store build?

I think Pico will fail formal builds, b/c  we can't upload the same version number to both channel 1  and channel 2. That's why we had that suffix.

I think that we specifically will need all of those FORCE_* for the Quest build; back when Q1 was still accepted, we needed them after we removed the OVR package. Now that we've removed it from Q2+ as well, I think we'll need them.

mikeskydev committed:
Use noncn region for all pico uploads

https://developer.picoxr.com/document/unity/command-line-utility/#3a500c65

mikeage committed:
Remove unneeded filter for Pico

mikeage committed:
Restore Quest-specific version to Github release

mikeage committed:
Use Quest specific build for ArborXR

mikeskydev commented:
Good catch on the symbols, but do we really need the quest specific version? It's just a package ID change... I'd rather use our proper ID on places we have control over

mikeskydev commented:
Oof, yeah I forgot about that... what a pain. I hate past me.

mikeskydev committed:
Reassign pico prefabs

I thought this was done?

mikeskydev committed:
update openxr extensions

mikeskydev committed:
and again

mikeskydev committed:
update again

pico pls

mikeskydev committed:
Merge branch 'main' into feature/pure-openxr

mikeage commented:
I looked at the manifests between the current Q2+ version and this Android Quest build. There are a few subtle differences:
This one adds `<uses-permission android:name="android.permission.BLUETOOTH"/>`
This one removes `<uses-permission android:name="com.oculus.permission.USE_SCENE"/>`, `<uses-permission android:name="com.oculus.permission.USE_ANCHOR_API"/>` and `<uses-permission android:name="com.oculus.permission.IMPORT_EXPORT_IOT_MAP_DATA"/>`
This one adds `<uses-permission android:name="com.oculus.permission.BOUNDARY_VISIBILITY"/>`. I think this was problematic before.

It also changes:
```
        <meta-data android:name="com.oculus.supportedDevices" android:value="quest2|questpro|eureka"/>
```
to
```
        <meta-data android:name="com.oculus.supportedDevices" android:value="quest|quest2"/>
```

There are a few other minor changes that I don't believe are significant. But these might be relevant (maybe not the permissions)

andybak committed:
Merge remote-tracking branch 'origin/main' into feature/pure-openxr

# Conflicts:
#	Assets/Scripts/Util/AndroidUtils.cs

andybak committed:
Merge remote-tracking branch 'origin/main' into feature/pure-openxr

# Conflicts:
#	.github/workflows/build.yml

andybak committed:
Post-merge fix

andybak committed:
Tweak brush angle by 10 degrees for Pico

andybak committed:
Revert "Tweak brush angle by 10 degrees for Pico"

This reverts commit e7ddb10317aab3070a02b965963c5ca03fad56b3.

andybak committed:
offset test

andybak committed:
get the orientation correct first

andybak committed:
Rotate the other direction

andybak committed:
Scripting support for debugging controll model position

andybak committed:
Update UnityXRPicoPhoenixControls.prefab

andybak committed:
Pico isn't registering grip pressed. Try an explicit threshold

andybak committed:
Temp override IsGreatFirewalled to always return true

andybak committed:
Fix faulty logic from previous commit

mikeskydev committed:
Merge branch 'main' into feature/pure-openxr

# Conflicts:
#	.github/workflows/build.yml

andybak committed:
Merge branch 'main' into feature/pure-openxr

andybak committed:
Merge remote-tracking branch 'origin/main' into feature/pure-openxr

# Conflicts:
#	Packages/manifest.json

andybak committed:
Merge branch 'main' into feature/pure-openxr

mikeage committed:
[REVERT THIS] temporarily disable the custom basename for PR builds

mikeage reviewed (changes requested):
Do not merge without reverting 899d6ed88

andybak committed:
See if explicitly hiding the overlay fixes the flickering on 2nd run

andybak committed:
Start with postfx disabled to see if it fixes flickering

andybak committed:
Try removing explicit CompositionLayer from XRRig

andybak committed:
Revert "Try removing explicit CompositionLayer from XRRig"

This reverts commit 141e42520d05ba5815b04b82604efa6c1f359404.

andybak committed:
Another potential fix for the flickering

"the loading root is DontDestroyOnLoad, and its camera has m_TargetEye: 3 (Both). After Main loads, that leaves the loading XR camera alive at the same time as the main XR camera. On pure OpenXR, that is a credible
  cause for one-eye/alternate-frame flicker. The first-launch permission dialogs likely mask the timing."

andybak committed:
package updates
